gadgets/trace_dns/program.bpf.c

```diff
@@ -256,48 +277,59 @@ int ig_trace_dns(struct __sk_buff *skb)
 
 	// Initialize event here only after we know we're interested in this packet to avoid
 	// spending useless cycles.
-	__builtin_memset(&event, 0, sizeof(event));
+	bpf_map_update_elem(&tmp_events, &zero, &empty_event, BPF_NOEXIST);
+	event = bpf_map_lookup_elem(&tmp_events, &zero);
+	if (!event)
+		return 0;
+	bpf_map_delete_elem(&tmp_events, &zero);
```

Comment:
I don't get how this works:
- `bpf_map_update_elem(&tmp_events, &zero, &empty_event, BPF_NOEXIST);` will always fail as the **array** map always has a key 0 present. 
- `bpf_map_delete_elem(&tmp_events, &zero);`: should fail since it's not really possible to remove an element from an array map. 

I added some printks to this and the result was:

```bash 
           <...>-1345    [000] ..s21  4305.236563: bpf_trace_printk: bpf_map_update_elem: -17
           <...>-1345    [000] ..s21  4305.236567: bpf_trace_printk: bpf_map_delete_elem: -17
    avahi-daemon-1345    [003] ..s21  4337.238352: bpf_trace_printk: bpf_map_update_elem: -17
    avahi-daemon-1345    [003] ..s21  4337.238359: bpf_trace_printk: bpf_map_delete_elem: -17
    avahi-daemon-1345    [008] ..s21  4401.237624: bpf_trace_printk: bpf_map_update_elem: -17
```

I pushed a commit to this PR that should fix it. In order to avoid merging my own changes, I'll wait for other folk to review it.

Reply:
Ah, this is also present on the built-in version of this gadget: Opened #3830 to fix it.